### PR TITLE
feat(Dropdown): Remove dropdown mask

### PR DESCRIPTION
**Pull request by lpoulter**
**Affected components**
* Dropdown
* AlertDropdown
* Select
* Datepicker

**Background**
The current implementation adds a mask to the DOM when a dropdown is open. This mask captures click events outside of the dropdown and causes the dropdown to be closed. This caused an undesirable double click when navigating between form elements. We wish to remove the mask, removing a click from the users flow.

**Approach**
This PR has proved challenging and I've tried a few different approaches. The one i've settled on is using a custom HOC and disabling pointer events. I could get no other approach to fully meet all the requirements. 
The `onclickoutside` HOC is used to wrap the dropdown source. This adds an event listener to the DOM which is then used to trigger closing of the dropdown. It was necessary to disable clicks, by disabling pointer events, on the dropdown target when the dropdown source is visible. This is to prevent the reopening of the dropdown once closed by a click-away. The mask previously captured this click.

**Deployed Branch here**
https://5b28fca0c965927b31dc9b9e--nostalgic-curie-a20a02.netlify.com/docs/packages/axiom-components/dropdown/

### CODE REVIEW

**Comment by jdelight** [comment]
Nice one @lpoulter 👍 Just had a very quick look - I was expecting to see we were setting pointer events to none at some point... but I can't see it. Might be worth just having a technical explanation of how this works in the PR description. Especially as there's just the one commit.

**Comment by jdelight** [comment]
Actually ignore my last comment - I think it's just the pointer events bit I was wondering about :)

**Comment by lpoulter** [comment]
Is this what your looking for https://github.com/BrandwatchLtd/axiom/pull/577/files#diff-76879590576e94f3f91fa935294d0140R93
I think the Axiom style is just one commit.

File: package.json [hunk 1]
```diff
@@ -59,6 +59,7 @@
     "react-addons-test-utils": "^16.0.0-0",
     "react-dom": "^16.2.0",
     "react-hot-loader": "^3.1.3",
+    "react-onclickoutside": "^6.7.1",
```
Comment (HHogg) [inline package.json]: This should be a dependency, not a dev dependency

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js [hunk 2]
```diff
@@ -1,15 +1,25 @@
 import React, { Component } from 'react';
 import PropTypes from 'prop-types';
 import Context from '../Context/Context';
+import onClickOutside from 'react-onclickoutside';
```
Comment (HHogg) [inline packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: We've always tried to stay pretty light on dependencies, and looking at this one it's been pretty quiet for a few months with a few issues around latest react support. 

Have you tried an approach without using a library to see how difficult it would be to achieve?

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
     type: PropTypes.oneOf(['success', 'warning', 'error', 'info']).isRequired,
+    closeDropdown: PropTypes.func.isRequired,
+  };
+
+  handleClickOutside = () => {
+    this.context.closeDropdown();
   };
 
   render() {
+    const onClickOutSideProps = ['eventTypes', 'outsideClickIgnoreClass', 'preventDefault', 'stopPropagation', 'disableOnClickOutside', 'enableOnClickOutside'];
```
Comment (HHogg) [inline packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: Omitting an external libraries list of props will take a bit of effort to maintain.

File: packages/axiom-components/src/AlertDropdown/AlertDropdown.js [hunk 4]
```diff
@@ -85,9 +85,11 @@ class AlertDropdown extends Component {
       <Position
           { ...rest }
           isVisible={ isVisible }
-          onMaskClick={ () => this.close() }
           position="bottom">
-        <PositionTarget>{ findComponent(children, AlertDropdownTargetRef) }</PositionTarget>
+        <PositionTarget>
+          { <span style={ { pointerEvents: isVisible ? 'none' : 'auto' } } >
```
Comment (HHogg) [inline packages/axiom-components/src/AlertDropdown/AlertDropdown.js]: This `span` will cause issues with layouts as it will introduce an element that will break the margin spacing pattern that is on every Axiom element.

File: package.json (hunk 1 above)
Comment (lpoulter) [reply package.json]: Ahh yes I wondered about that. I can't see a dependency section in the package.json are they added in some other way?

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js (hunk 2 above)
Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: This library essentially adds events handlers to the body when the wrapped component is mounted. It also handles some of the ugly ref passing/getting and provides some useful options. It has pretty decent tests also. Rather than re-writing  a solution that works well. I think it's worth using The 16.3 issues don't look like problems to me more suggested refactors to make use of the new features in 16.3.

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js (hunk 3 above)
Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: Yes it is ugly. I'll see if I can find a better way.

File: packages/axiom-components/src/AlertDropdown/AlertDropdown.js (hunk 4 above)
Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdown.js]: Hmm this is a tricky one. I could clone the child and add a `pointerEventsDisabled` prop to it.  However each possible child of `AlertDropdownTarget` and `DropdownTarget` would need to know what to do with it. Is there anyway I can keep the `pointerEvents` logic here and not affect the DOM structure.

**Comment by lpoulter** [comment]
@HHogg I've tried a few different approaches both with and without the library and a HOC seems like the best approach. Is there any particular reason you've avoid HOC's in Axiom and could you suggest any alternative patterns?

**Comment by jdelight** [comment]
@lpoulter / @HHogg I'm not opposed to using a library if there is one that's well written and not bloated with features we don't need (at the moment). 

But it would be worth taking some time to evaluate the available libraries, determine why we would choose one over the others (there are 3 I can see on a after a quick Google: https://github.com/kentor/react-click-outside, https://github.com/tj/react-click-outside, https://github.com/Pomax/react-onclickoutside).

Have you seen this one from TJ: https://github.com/tj/react-click-outside/blob/master/index.js looks very small / simple etc... does it work for what we need? If so is there value in writing one ourselves over using it directly. Maybe our version could be even simpler and use it as inspiration.

I think this is one of those features which doesn't typically need updating lots so I'm not too concerned about activity on the repo if it's "complete". But I agree with @HHogg that it's good to be extra cautious about which dependencies we introduce.

**Comment by lpoulter** [comment]
Thanks for the links James. I've looked at these solutions. The simpler libraries such as TJ's library don't provide a reference to the wrapped component instance. This instance is used to call closeDropdown() within the clickOutside handler and allows for a more decoupled implementation.

The kentor and Pomax versions use a similar method, using the wrapped component instance. However I had some problems implementing the Kentor solution. The Pomax version is also used more widely and unlike kentor adds no transitive dependencies.

It's definitely possible to write our own solution, ideally using the new `React.createRef()` added in `16.3`, axiom is `16.2`.

**Comment by jdelight** [comment]
Very interesting - thanks @lpoulter

**Comment by lpoulter** [comment]
So this might be enough to cover our use case. Will need to add some edge cases and test. https://codesandbox.io/s/o4lr9z7l8y

**Comment by lpoulter** [comment]
@jdelight @HHogg I've added a commit showing the use of my own implementation of an `onClickOutside` are you able to comment on the approach?

File: packages/axiom-components/src/Base/Base.js [hunk 5]
```diff
@@ -34,6 +34,8 @@ export default class Base extends Component {
      * Opposite of `visibleUntil`.
      */
     hiddenUntil: PropTypes.oneOf(['small', 'medium', 'large']),
+    /** disables pointer events */
+    pointerEventsDisabled: PropTypes.bool,
```
Comment (HHogg) [inline packages/axiom-components/src/Base/Base.js]: What's the reason for disabling pointer events?

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js [hunk 6]
```diff
@@ -13,3 +19,5 @@ export default class AlertDropdownContext extends Component {
     );
   }
 }
+
+export default withOnClickOutside(AlertDropdownContext);
```
Comment (HHogg) [inline packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: How come we're wrapping each use of Dropdown rather than add the behaviour to the Dropdown component?

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js [hunk 7]
```diff
@@ -1,10 +1,16 @@
...
+import { withOnClickOutside } from '@brandwatch/axiom-utils';
 
-export default class AlertDropdownContext extends Component {
+class AlertDropdownContext extends Component {
   static contextTypes = {
     type: PropTypes.oneOf(['success', 'warning', 'error', 'info']).isRequired,
+    closeDropdown: PropTypes.func.isRequired,
+  };
+
+  handleClickOutside = () => {
```
Comment (HHogg) [inline packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: The arrow method syntax was decided against (I can't find the PR) a while ago.

File: packages/axiom-components/src/Base/Base.js (hunk 5 above)
Comment (lpoulter) [reply packages/axiom-components/src/Base/Base.js]: Explained in the PR desc.

 "It was nessacary to disable clicks, by disabling pointer events, on the dropdown target when the dropdown source is visible. This is to the reopening of the dropdown once closed by a click-away."

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js (hunk 7 above)
Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: Any reason why. I like using the arrow syntax to avoid having to bind this in every constructor.

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js (hunk 6 above)
Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: Putting the logic in a HOC allows for it to be used in multiple components. Currently both `Dropdown` and `AlertDropdown` need this functionality.

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js (hunk 7 above)
Comment (HHogg) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: IIRC it was to do with the spec stage it was in... personally I don't mind it. 

Axiom has always lent towards pattern consistency, so would you mind opening up a discussion to see how others feel and if everyone's happy to to do a full switch over to this pattern?

Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: Sure will do. In the mean time i'll use the standard syntax.

File: packages/axiom-components/src/Base/Base.js (hunk 5 above)
Comment (HHogg) [reply packages/axiom-components/src/Base/Base.js]: Is this likely to cause issues for elements that still need some sort of mouse interaction (like a hover tooltip)? It just seems like a catch all approach, when the goal is to not reopen the dropdown.

It might be safer to try an approach of disabling the open dropdown method when it's already open or stop the event propagating up (when it's open) to the handler on the document.

Comment (lpoulter) [reply packages/axiom-components/src/Base/Base.js]: "It might be safer to try an approach of disabling the open dropdown method when it's already open or stop the event propagating up (when it's open) to the handler on the document."

I did try these first, I agree they are better options. I can have another go with now i'm using my own implementation.

Comment (lpoulter) [reply packages/axiom-components/src/Base/Base.js]: "Is this likely to cause issues for elements that still need some sort of mouse interaction (like a hover tooltip)?"

I imagine the current mask implementation has this issue also?

Comment (jdelight) [reply packages/axiom-components/src/Base/Base.js]: I was thinking the same @lpoulter - presumably the `pointer-events: none` is only applicable to the dropdown target (i.e. the thing you initially click on). So all other hover / tooltip interactions should work as normal.

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js (hunk 6 above)
Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: We also want to wrap the `context` components rather than the whole dropwdown components. That way the event listener are only added when the dropdown is open.

File: packages/axiom-components/src/Base/Base.js (hunk 5 above)
Comment (HHogg) [reply packages/axiom-components/src/Base/Base.js]: Sorry I meant like a tooltip/title inside the target element. Yes the current mask does prevent this, but if we're removing the mask so interactions can happen while the dropdown is open, then it might throw people of as to why the interactions on the target don't work. 

If the behaviour to stop it reopening on clicking the target can be isolated to just that rather than stopping all interactions on the target then it's just less likely to trip up someone down the line.

File: packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js (hunk 6 above)
Comment (HHogg) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: I wouldn't worry to much about `AlertDropdown` as it was decided to remove that component as it doesn't fit well into the design system, and the adding and removing of event listeners can still happen in the `open` and `close` methods of Dropdown. 

It's just this way it keeps them more inline as a pure react component and we can limit the implementation inside Dropdown (ie not requiring components to implement a `handleClickOutside`).

Comment (lpoulter) [reply packages/axiom-components/src/AlertDropdown/AlertDropdownContext.js]: I can move all the logic to the `Dropdown` component then. We will lose the ability to add `handleClickOutside` functionality to other components but maybe thats ok.

File: packages/axiom-components/src/Base/Base.js (hunk 5 above)
Comment (jdelight) [reply packages/axiom-components/src/Base/Base.js]: Sorry @HHogg not sure I follow :/ do you have an example of this on Axiom or in Audiences? thanks

**Comment by jdelight** [comment]
@lpoulter the first version (commit) looks a lot better to me! what do you think? much less imperative state management and using event bubbling / propagation is much more how I would expect this to be done.

I think the 2 things @HHogg was after can still be achieved i.e. only applying it to the Dropdown component & not using HOC (I'm not as fussed about HOC but would be interested to see how it would work without using one). 

Perhaps if we can do that we're on to a winner with something like the first commit. 🎉 

@siggysamson / @thatguynamedandy what do you think?

**Comment by lpoulter** [comment]
@thatguynamedandy Would be great to get your input on this now harry is away. I've gone with the HOC approach as the second approach had a bug and the need for extra state to be track.

**Comment by jdelight** [comment]
Thanks for taking a look @thatguynamedandy 👍 @lpoulter what do you think?

**Comment by lpoulter** [comment]
Thanks for taking a look @thatguynamedandy. I've already tried a solution like yours there are a couple of problems with it. Ignoring the click on the dropdown target means that we lose the toggle behaviour, clicking on target when open should close the dropdown.

HOC's are a widely used react pattern with many article and even the offical docs discussing them at length. The benefit of using one here is that the code is re-useable, allowing me to reuse the `clickOutside` functionality in any other component that may need it, both `Alertdropdown` and any future ones. However I understand your point about not wanting to many patterns in Axiom. I've updated my soultion to not use a HOC.

Are you able to take another look?
